File: occupancy-manager-java/src/main/java/com/occupancy/manager/service/OccupancyManagerService.java
package com.occupancy.manager.service;

import static java.util.Collections.sort;
import static java.util.stream.Collectors.partitioningBy;
import static java.util.stream.Collectors.toList;

import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import com.occupancy.manager.dto.OccupancyUsageRequest;
import com.occupancy.manager.dto.OccupancyUsageResponse;
import com.occupancy.manager.dto.OccupancyDataStore;

import lombok.extern.slf4j.Slf4j;

/**
 * 
 * @author Ramachandra
 *
 */
@Slf4j
@Service
public class OccupancyManagerService implements IOccupancyManagerService {

	private static final String OCCUPANCY_USAGE = "%s (EUR %s)";

	public OccupancyUsageResponse calculateOccupancy(OccupancyUsageRequest occupancyUsageRequest) {

		var premiumRooms = occupancyUsageRequest.getPremiumRooms();
		var economyRooms = occupancyUsageRequest.getEconomyRooms();

		log.info("Premium Rooms {}", premiumRooms);
		log.info("Economy Rooms {}", economyRooms);

		var partitionCustomers = getPartitionCustomers(occupancyUsageRequest);
		var economyCustomers = partitionCustomers.get(0);
		var premiumCustomers = partitionCustomers.get(1);

		log.info("Economy Customers {}", economyCustomers);
		log.info("Premium Customers {}", premiumCustomers);

		var occupancyDataStore = new OccupancyDataStore();

		calculatePremiumCustomersUsage(premiumCustomers, premiumRooms, occupancyDataStore);

		calculateEconomyCustomersWithPremiumUpgrade(economyCustomers, economyRooms, occupancyDataStore);

		log.info("Usage Premium: {} (EUR {})", occupancyDataStore.getPremiumRoomsOccupied(),
				occupancyDataStore.getPremiumUsage());
		log.info("Usage Economy: {} (EUR {})", occupancyDataStore.getEconomyRoomsOccupied(),
				occupancyDataStore.getEconomyUsage());
		return new OccupancyUsageResponse(
				String.format(OCCUPANCY_USAGE, occupancyDataStore.getPremiumRoomsOccupied(),
						occupancyDataStore.getPremiumUsage()),
				String.format(OCCUPANCY_USAGE, occupancyDataStore.getEconomyRoomsOccupied(),
						occupancyDataStore.getEconomyUsage()));
	}

	/**
	 * This method will sort and provides the partition customers by margin
	 * 
	 * @param occupancyUsageRequest
	 * @return
	 */
	private List<List<Integer>> getPartitionCustomers(OccupancyUsageRequest occupancyUsageRequest) {
		var customerPayments = occupancyUsageRequest.getPayments();
		sort(customerPayments);
		Predicate<Integer> minimumPremiumCondition = value -> value >= occupancyUsageRequest.getMinimumPremium();
		return partitionCustomersByMinimumPremium(customerPayments, minimumPremiumCondition);
	}

	/**
	 * This method is to calculate the sum of all list
	 * 
	 * @param customers
	 * @return sum
	 */
	private int calculateSum(List<Integer> customers) {
		return customers.stream().mapToInt(Integer::intValue).sum();
	}

	/**
	 * This method is the calculate the usage of the premium customers
	 * 
	 * @param premiumCustomers
	 * @param premiumRooms
	 * @param occupancyDataStore
	 */
	private void calculatePremiumCustomersUsage(List<Integer> premiumCustomers, int premiumRooms,
			OccupancyDataStore occupancyDataStore) {

		if (premiumRooms > 0 && !premiumCustomers.isEmpty()) {
			var totalPremiumCustomers = premiumCustomers.size();
			if (premiumRooms >= totalPremiumCustomers) {
				occupancyDataStore.setPremiumUsage(
						calculateSum(occupancyDataStore.getPremiumUsage(), calculateSum(premiumCustomers)));
				occupancyDataStore.setPremiumRoomsVacant(
						calculateSum(occupancyDataStore.getPremiumRoomsVacant(), premiumRooms - totalPremiumCustomers));
				occupancyDataStore.setPremiumRoomsOccupied(
						calculateSum(occupancyDataStore.getPremiumRoomsOccupied(), totalPremiumCustomers));
			} else {
				var premiumCustomersOccupied = getSubListCustomers(premiumCustomers,
						totalPremiumCustomers - premiumRooms, totalPremiumCustomers);
				occupancyDataStore.setPremiumUsage(
						calculateSum(occupancyDataStore.getPremiumUsage(), calculateSum(premiumCustomersOccupied)));
				occupancyDataStore.setPremiumRoomsOccupied(
						calculateSum(occupancyDataStore.getPremiumRoomsOccupied(), premiumCustomersOccupied.size()));
			}
		}
	}

	/**
	 * 
	 * @param economyCustomers
	 * @param economyRooms
	 * @param occupancyDataStore
	 */
	private void calculateEconomyCustomersWithPremiumUpgrade(List<Integer> economyCustomers, Integer economyRooms,
			OccupancyDataStore occupancyDataStore) {
		// Validating the upgrade of the economy customers
		if (economyRooms > 0 && !economyCustomers.isEmpty()) {
			var totalEconomyCustomers = economyCustomers.size();
			// Check whether the economy customers are prefilled even if the customers are
			// upgraded to premium
			if (totalEconomyCustomers >= (economyRooms - 1) && occupancyDataStore.getPremiumRoomsVacant() > 0) {
				var economyCustomersOccupied = economyRooms >= totalEconomyCustomers
						? getSubListCustomers(economyCustomers, 0, totalEconomyCustomers)
						: getSubListCustomers(economyCustomers, 0, totalEconomyCustomers - economyRooms);
				var economyPremiumOccupied = economyCustomers.stream().collect(toList());
				economyPremiumOccupied.removeAll(economyCustomersOccupied);
				if (!economyPremiumOccupied.isEmpty()) {
					occupancyDataStore.setPremiumUsage(
							calculateSum(occupancyDataStore.getPremiumUsage(), calculateSum(economyPremiumOccupied)));
					occupancyDataStore.setPremiumRoomsOccupied(
							calculateSum(occupancyDataStore.getPremiumRoomsOccupied(), economyPremiumOccupied.size()));
				}
				calculateEconomyCustomersUsage(economyCustomersOccupied, economyRooms, occupancyDataStore);
			} else {
				calculateEconomyCustomersUsage(economyCustomers, economyRooms, occupancyDataStore);
			}
		}
	}

	/**
	 * This method is the calculate the usage of the economy customers
	 * 
	 * @param economyCustomers
	 * @param economyRooms
	 * @param occupancyDataStore
	 */
	private void calculateEconomyCustomersUsage(List<Integer> economyCustomers, int economyRooms,
			OccupancyDataStore occupancyDataStore) {
		var totalEconomyCustomers = economyCustomers.size();
		if (economyRooms >= totalEconomyCustomers) {
			occupancyDataStore
					.setEconomyUsage(calculateSum(occupancyDataStore.getEconomyUsage(), calculateSum(economyCustomers)));
			occupancyDataStore.setEconomyRoomsOccupied(
					calculateSum(occupancyDataStore.getEconomyRoomsOccupied(), totalEconomyCustomers));
		} else {
			var economyCustomersOccupied = getSubListCustomers(economyCustomers, totalEconomyCustomers - economyRooms,
					totalEconomyCustomers);
			occupancyDataStore.setEconomyUsage(
					calculateSum(occupancyDataStore.getEconomyUsage(), calculateSum(economyCustomersOccupied)));
			occupancyDataStore.setEconomyRoomsOccupied(
					calculateSum(occupancyDataStore.getEconomyRoomsOccupied(), economyCustomersOccupied.size()));
		}
	}

	/**
	 * This method is to calculate the sum of two values
	 * 
	 * @param initialData
	 * @param updatedData
	 * @return sum
	 */
	private int calculateSum(int initialData, int updatedData) {
		return initialData + updatedData;
	}

	/**
	 * This method is to build the sublist of the customers
	 * 
	 * @param customers
	 * @param startIndex
	 * @param lastIndex
	 * @return sublist
	 */
	private List<Integer> getSubListCustomers(List<Integer> customers, int startIndex, int lastIndex) {
		return customers.stream().skip(startIndex).limit(lastIndex).collect(toList());
	}

	/**
	 * This method will create two list of customers based on the predicate
	 * 
	 * @param customers
	 * @param condition
	 * @return list of customers
	 */
	private <T> List<List<T>> partitionCustomersByMinimumPremium(List<T> customers, Predicate<T> condition) {
		return customers.stream().collect(partitioningBy(s -> (condition.test(s)))).values().stream()
				.collect(Collectors.toList());
	}
}

File: occupancy-manager-kotlin/src/main/kotlin/com/occupancy/manager/service/OccupancyManagerService.kt
package com.occupancy.manager.service

import com.occupancy.manager.dto.OccupancyDataStore
import com.occupancy.manager.dto.OccupancyUsageRequest
import com.occupancy.manager.dto.OccupancyUsageResponse
import mu.KotlinLogging
import org.springframework.stereotype.Service
import java.util.Collections.sort
import java.util.function.Predicate
import java.util.stream.Collectors

/**
 *
 * @author Ramachandra
 */
internal val logger = KotlinLogging.logger {}
@Service
class OccupancyManagerService : IOccupancyManagerService {
    override fun calculateOccupancy(occupancyUsageRequest: OccupancyUsageRequest): OccupancyUsageResponse {
        val premiumRooms: Int = occupancyUsageRequest.premiumRooms
        val economyRooms: Int = occupancyUsageRequest.economyRooms
        logger.info("Premium Rooms {}", premiumRooms)
        logger.info("Economy Rooms {}", economyRooms)
        val partitionCustomers = getPartitionCustomers(occupancyUsageRequest)
        val economyCustomers = partitionCustomers[0]
        val premiumCustomers = partitionCustomers[1]
        logger.info("Economy Customers {}", economyCustomers)
        logger.info("Premium Customers {}", premiumCustomers)
        val occupancyDataStore = OccupancyDataStore()
        calculatePremiumCustomersUsage(premiumCustomers, premiumRooms, occupancyDataStore)
        calculateEconomyCustomersWithPremiumUpgrade(economyCustomers, economyRooms, occupancyDataStore)
        logger.info(
            "Usage Premium: {} (EUR {})", occupancyDataStore.premiumRoomsOccupied,
            occupancyDataStore.premiumUsage
        )
        logger.info(
            "Usage Economy: {} (EUR {})", occupancyDataStore.economyRoomsOccupied,
            occupancyDataStore.economyUsage
        )
        return OccupancyUsageResponse(
            String.format(
                OCCUPANCY_USAGE, occupancyDataStore.premiumRoomsOccupied,
                occupancyDataStore.premiumUsage
            ), String.format(
                OCCUPANCY_USAGE, occupancyDataStore.economyRoomsOccupied,
                occupancyDataStore.economyUsage
            )
        )
    }

    /**
     * This method will sort and provides the partition customers by margin
     *
     * @param occupancyUsageRequest
     * @return
     */
    private fun getPartitionCustomers(occupancyUsageRequest: OccupancyUsageRequest): List<List<Int>> {
        val customerPayments: List<Int> = occupancyUsageRequest.payments
        sort(customerPayments)
        val minimumPremiumCondition =
            Predicate { value: Int -> value >= occupancyUsageRequest.minimumPremium }
        return partitionCustomersByMinimumPremium(customerPayments, minimumPremiumCondition)
    }

    /**
     * This method is to calculate the sum of all list
     *
     * @param customers
     * @return sum
     */
    private fun calculateSum(customers: List<Int>): Int {
        return customers.stream().mapToInt { obj: Int -> obj }.sum()
    }

    /**
     * This method is the calculate the usage of the premium customers
     *
     * @param premiumCustomers
     * @param premiumRooms
     * @param occupancyDataStore
     */
    private fun calculatePremiumCustomersUsage(
        premiumCustomers: List<Int>, premiumRooms: Int,
        occupancyDataStore: OccupancyDataStore
    ) {
        if (premiumRooms > 0 && premiumCustomers.isNotEmpty()) {
            val totalPremiumCustomers = premiumCustomers.size
            if (premiumRooms >= totalPremiumCustomers) {
                occupancyDataStore.premiumUsage =
                    calculateSum(occupancyDataStore.premiumUsage, calculateSum(premiumCustomers))

                occupancyDataStore.premiumRoomsVacant =
                    calculateSum(occupancyDataStore.premiumRoomsVacant, premiumRooms - totalPremiumCustomers)

                occupancyDataStore.premiumRoomsOccupied=
                    calculateSum(occupancyDataStore.premiumRoomsOccupied, totalPremiumCustomers)

            } else {
                val premiumCustomersOccupied = getSubListCustomers(
                    premiumCustomers,
                    totalPremiumCustomers - premiumRooms, totalPremiumCustomers
                )
                occupancyDataStore.premiumUsage =
                    calculateSum(occupancyDataStore.premiumUsage, calculateSum(premiumCustomersOccupied))

                occupancyDataStore.premiumRoomsOccupied =
                    calculateSum(occupancyDataStore.premiumRoomsOccupied, premiumCustomersOccupied.size)

            }
        }
    }

    /**
     *
     * @param economyCustomers
     * @param economyRooms
     * @param occupancyDataStore
     */
    private fun calculateEconomyCustomersWithPremiumUpgrade(
        economyCustomers: List<Int>, economyRooms: Int,
        occupancyDataStore: OccupancyDataStore
    ) {
        // Validating the upgrade of the economy customers
        if (economyRooms > 0 && economyCustomers.isNotEmpty()) {
            val totalEconomyCustomers = economyCustomers.size
            // Check whether the economy customers are prefilled even if the customers are
            // upgraded to premium
            if (totalEconomyCustomers >= economyRooms - 1 && occupancyDataStore.premiumRoomsVacant > 0) {
                val economyCustomersOccupied = if (economyRooms >= totalEconomyCustomers) getSubListCustomers(
                    economyCustomers,
                    0,
                    totalEconomyCustomers
                ) else getSubListCustomers(economyCustomers, 0, totalEconomyCustomers - economyRooms)
                val economyPremiumOccupied = economyCustomers.stream().collect(Collectors.toList())
                economyPremiumOccupied.removeAll(economyCustomersOccupied)
                if (economyPremiumOccupied.isNotEmpty()) {
                    occupancyDataStore.premiumUsage =
                        calculateSum(occupancyDataStore.premiumUsage, calculateSum(economyPremiumOccupied))

                    occupancyDataStore.premiumRoomsOccupied =
                        calculateSum(occupancyDataStore.premiumRoomsOccupied, economyPremiumOccupied.size)

                }
                calculateEconomyCustomersUsage(economyCustomersOccupied, economyRooms, occupancyDataStore)
            } else {
                calculateEconomyCustomersUsage(economyCustomers, economyRooms, occupancyDataStore)
            }
        }
    }

    /**
     * This method is the calculate the usage of the economy customers
     *
     * @param economyCustomers
     * @param economyRooms
     * @param occupancyDataStore
     */
    private fun calculateEconomyCustomersUsage(
        economyCustomers: List<Int>, economyRooms: Int,
        occupancyDataStore: OccupancyDataStore
    ) {
        val totalEconomyCustomers = economyCustomers.size
        if (economyRooms >= totalEconomyCustomers) {
            occupancyDataStore
                .economyUsage=calculateSum(occupancyDataStore.economyUsage, calculateSum(economyCustomers))
            occupancyDataStore.economyRoomsOccupied =
                calculateSum(occupancyDataStore.economyRoomsOccupied, totalEconomyCustomers)

        } else {
            val economyCustomersOccupied = getSubListCustomers(
                economyCustomers, totalEconomyCustomers - economyRooms,
                totalEconomyCustomers
            )
            occupancyDataStore.economyUsage =
                calculateSum(occupancyDataStore.economyUsage, calculateSum(economyCustomersOccupied))
            
            occupancyDataStore.economyRoomsOccupied =
                calculateSum(occupancyDataStore.economyRoomsOccupied, economyCustomersOccupied.size)

        }
    }

    /**
     * This method is to calculate the sum of two values
     *
     * @param initialData
     * @param updatedData
     * @return sum
     */
    private fun calculateSum(initialData: Int, updatedData: Int): Int {
        return initialData + updatedData
    }

    /**
     * This method is to build the sublist of the customers
     *
     * @param customers
     * @param startIndex
     * @param lastIndex
     * @return sublist
     */
    private fun getSubListCustomers(customers: List<Int>, startIndex: Int, lastIndex: Int): List<Int> {
        return customers.stream().skip(startIndex.toLong()).limit(lastIndex.toLong()).collect(Collectors.toList())
    }

    /**
     * This method will create two list of customers based on the predicate
     *
     * @param customers
     * @param condition
     * @return list of customers
     */
    private fun <T> partitionCustomersByMinimumPremium(customers: List<T>, condition: Predicate<T>): List<List<T>> {
        return customers.stream().collect(Collectors.partitioningBy { s: T ->
            condition.test(
                s
            )
        }).values.stream()
            .collect(Collectors.toList())
    }

    companion object {
        private const val OCCUPANCY_USAGE = "%s (EUR %s)"
    }
}